kernel/mellanox_ethernet/src/command_queue.rs

```diff
@@ -0,0 +1,740 @@
+/// It resides in a physically contiguous 4 KiB memory chunk.
+/// (Section 8.24.1: HCA Command Queue)
+#[repr(C)]
+pub struct CommandQueue {
+    /// Physically-contiguous command queue entries
+    entries: BoxRefMut<MappedPages, [CommandQueueEntry]>,
+    /// Per-entry boolean flags to keep track of which entries are in use
+    available_entries: Vec<bool>,
+    /// A random number that needs to be different for every command, and the same for all mailboxes that are part of a command.
+    token: u8,
+    /// Pages used for input mailboxes 
+    mailbox_buffers_input: Vec<Vec<(MappedPages, PhysicalAddress)>>, 
+    /// Pages used for output mailboxes 
+    mailbox_buffers_output: Vec<Vec<(MappedPages, PhysicalAddress)>> 
+}
+
+impl CommandQueue {
+
+    /// Create a command queue object.
+    ///
+    /// # Arguments
+    /// * `entries`: physically contiguous memory that is mapped as a slice of command queue entries.
+    /// * `num_cmdq_entries`: number of entries in the queue.
+    pub fn create(entries: BoxRefMut<MappedPages, [CommandQueueEntry]>, num_cmdq_entries: usize) -> Result<CommandQueue, &'static str> {
+        
+        // initially, all command entries are available
+        let mut available_entries = Vec::with_capacity(num_cmdq_entries);
+        for _ in 0..num_cmdq_entries { available_entries.push(true); }
+
+        // start off by pre-allocating one page for input and output mailboxes per entry
+        let mut mailbox_buffers_input = Vec::with_capacity(num_cmdq_entries);
+        let mut mailbox_buffers_output = Vec::with_capacity(num_cmdq_entries);
+        for _ in 0..num_cmdq_entries {
+            let (mailbox_mp, mailbox_pa) = create_contiguous_mapping(PAGE_SIZE, NIC_MAPPING_FLAGS)?;
+            mailbox_buffers_input.push(vec!((mailbox_mp, mailbox_pa)));
+
+            let (mailbox_mp, mailbox_pa) = create_contiguous_mapping(PAGE_SIZE, NIC_MAPPING_FLAGS)?;
+            mailbox_buffers_output.push(vec!((mailbox_mp, mailbox_pa)));
+        }
+
+        // Assign a random number to token, another OS (Snabb) starts off with 0xAA
+        let token = 0xAA;
+
+        Ok(CommandQueue{ entries, available_entries, token, mailbox_buffers_input, mailbox_buffers_output })
+    }
+
+    /// Find an command queue entry that is not in use
+    fn find_free_command_entry(&self) -> Option<usize> {
+        self.available_entries.iter().position(|&x| x == true)
+    }
+
+    /// Fill in the fields of a command queue entry.
+    /// At the end of the function, the command is ready to be posted using the doorbell in the initialization segment. 
+    /// Returns an error if no entry is available to use.
+    ///
+    /// # Arguments
+    /// * `opcode`: opcode for command that the driver wants to execute
+    /// * `opmod`: opcode modifer, only applicable for certain commands
+    /// * `allocated_pages`: physical address of pages that need to be passed to the NIC. Only used in the MANAGE_PAGES command.
```

Comment:
doc comment here should link to the `ManagePages` variant of the `CommandOpcode` enum